### PR TITLE
Detect table scans

### PR DESCRIPTION
**Pull request by tirsen**

### CODE REVIEW

File: misk-hibernate-testing/src/main/kotlin/misk/hibernate/HibernateTestingModule.kt [hunk 1]
```diff
@@ -26,34 +27,37 @@ class HibernateTestingModule(
...
+
     val transacterKey = Transacter::class.toKey(qualifier)
     val transacterProvider = getProvider(transacterKey)
 
-    if ((config == null || config.type == DataSourceType.VITESS)
-      && !disableCrossShardQueryDetector) {
+    if ((config == null || config.type == DataSourceType.VITESS) && !disableChecks) {
       bindVitessChecks()
+    } else {
+      val scaleSafetyChecksKey = ScaleSafetyChecks::class.toKey(qualifier)
```
Comment (swankjesse) [inline misk-hibernate-testing/src/main/kotlin/misk/hibernate/HibernateTestingModule.kt]: Isn't this checksKey?

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt [hunk 2]
```diff
@@ -46,6 +53,188 @@ internal data class Variables(
   val QueriesProcessed: Map<String, Int>
 )
 
+/**
+ * A MySQL query explanation. Ported from com.squareup.franklin.sqlcheck.MySqlQueryChecker in the
```
Comment (swankjesse) [inline misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt]: FWIW, origin story in open source is weird

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt [hunk 3]
```diff
@@ -46,6 +53,188 @@ internal data class Variables(
...
+   * the table condition. That is, rows shows the estimated number of rows examined and
+   * rows × filtered / 100 shows the number of rows that will be joined with previous tables.
+   */
+  internal var filtered: Double? = null
+  /**
+   * This column contains additional information about how MySQL resolves the query. For
+   * descriptions of the different values, see
+   * https://dev.mysql.com/doc/refman/5.7/en/explain-output.html#explain-extra-information.
+   */
+  internal var Extra: String? = null
```
Comment (swankjesse) [inline misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt]: The uppercase here is surprising... Binding or something?

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt [hunk 4]
```diff
@@ -46,6 +53,188 @@ internal data class Variables(
...
+        // See https://dev.mysql.com/doc/refman/5.7/en/subquery-materialization.html.
+        return true
+      }
+    }
+
+    return "PRIMARY" == key && query.contains(" limit ")
+  }
+
+  override fun toString(): String {
+    val builder = StringBuilder()
```
Comment (swankjesse) [inline misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt]: You might like buildString.

Search the codebase for an example.

File: misk-hibernate/src/main/kotlin/misk/jdbc/ScaleSafetyChecks.kt [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package misk.jdbc
+
+interface ScaleSafetyChecks {
+  fun <T> disable(body: () -> T): T
```
Comment (swankjesse) [inline misk-hibernate/src/main/kotlin/misk/jdbc/ScaleSafetyChecks.kt]: I don't think this API is obvious... Is it just “runs body if checks are disabled" ?

Comment (swankjesse) [reply misk-hibernate/src/main/kotlin/misk/jdbc/ScaleSafetyChecks.kt]: Nope, its “runs body with checks disabled” !

I think this probably wants to be a offered via an on Session? That's already handy when you're gonna wanna disable checks right?

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt (hunk 3 above)
Comment (swankjesse) [reply misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt]: Yes, line 224 shows the goods. Cool!

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt [hunk 6]
```diff
@@ -208,6 +401,74 @@ class VitessScaleSafetyChecks(
...
+          val explanations = c.createStatement().use { s ->
+            potentialKeyspaces.map(fun(keyspace): List<Explanation>? {
+              val database = if (keyspace.value.sharded) {
+                "vt_${keyspace.key}_-80"
+              } else {
+                "vt_${keyspace.key}_0"
+              }
+              s.execute("USE `$database`")
+              return try {
+                println(rawQuery)
```
Comment (swankjesse) [inline misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt]: Remove?!

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt (hunk 4 above)
Comment (tirsen) [reply misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt]: Ohhhh coooool!

File: misk-hibernate/src/main/kotlin/misk/jdbc/ScaleSafetyChecks.kt (hunk 5 above)
Comment (tirsen) [reply misk-hibernate/src/main/kotlin/misk/jdbc/ScaleSafetyChecks.kt]: Done! I like it! My main gripe is that now I have to provide a `Transacter::currentSession` method which I'm not too happy about. Do you have other suggestions?

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/TruncateTablesService.kt [hunk 7]
```diff
@@ -25,70 +25,60 @@ class TruncateTablesService(
...
-        val tableNamesQuery = when (config.type) {
-          DataSourceType.MYSQL -> {
-            "SELECT table_name FROM information_schema.tables where table_schema='${config.database}'"
-          }
-          DataSourceType.HSQLDB -> {
-            "SELECT TABLE_NAME FROM INFORMATION_SCHEMA.SYSTEM_TABLES WHERE TABLE_TYPE='TABLE'"
-          }
-          DataSourceType.VITESS -> {
-            "SHOW VSCHEMA TABLES"
+        session.withoutChecks {
```
Comment (swankjesse) [inline misk-hibernate-testing/src/main/kotlin/misk/jdbc/TruncateTablesService.kt]: Nice, I like how this reads!

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt [hunk 8]
```diff
@@ -46,6 +54,186 @@ internal data class Variables(
   val QueriesProcessed: Map<String, Int>
 )
 
+/**
+ * A MySQL query explanation.
+ */
+internal class Explanation {
```
Comment (swankjesse) [inline misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt]: The filename is `VitessScaleSafetyChecks` but I think this works on MySQL too?

File: misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt [hunk 9]
```diff
@@ -336,8 +345,14 @@ internal class RealTransacter private constructor(
...
     }
 
+    override fun <T> withoutChecks(body: () -> T): T = checksEnabled.withValue(false, body)
+
+    override fun areChecksEnabled(): Boolean = checksEnabled.get()
+
     companion object {
       private val log = getLogger<RealSession>()
+
+      private val checksEnabled = ThreadLocal.withInitial { true }
```
Comment (swankjesse) [inline misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt]: Move this val to be defined on the Transacter, not on the Session?

File: misk-hibernate/src/main/kotlin/misk/hibernate/Transacter.kt [hunk 10]
```diff
@@ -9,6 +9,11 @@ interface Transacter {
    */
   val inTransaction: Boolean
 
+  /**
+   * Returns the session of the currently running transaction.
+   */
+  val currentSession: Session
```
Comment (swankjesse) [inline misk-hibernate/src/main/kotlin/misk/hibernate/Transacter.kt]: Yuck, yeah if we can do without this that’d be good

Comment (swankjesse) [reply misk-hibernate/src/main/kotlin/misk/hibernate/Transacter.kt]: I’d prefer `areChecksEnabled()` here.

File: misk-hibernate/src/main/kotlin/misk/jdbc/ScaleSafetyChecks.kt (hunk 5 above)
Comment (swankjesse) [reply misk-hibernate/src/main/kotlin/misk/jdbc/ScaleSafetyChecks.kt]: Yes! The API goes on Session but it manipulates a property of the transacter

File: misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt (hunk 8 above)
Comment (tirsen) [reply misk-hibernate-testing/src/main/kotlin/misk/jdbc/VitessScaleSafetyChecks.kt]: Yes in theory but the currently implementation is coupled to the way MySQL runs when we're using Vitess.

**Pull request merged**